### PR TITLE
Reader: add PostComment to devdocs w/ new classNames

### PR DESCRIPTION
**Pull request by samouri**
This PR adds the `PostComment` class to our devdocs so that we can parallelize styling efforts with the rest of implementation.  It also adds in a new prop `displayType` with three possible values:

|key |css class | meaning|
|---- |---|---|
|full | `is-full` | default behavior. show the full comment|
|excerpt | `is-excerpt` | default behavior for convo tool.  show 3 lines by default w/ an option for read more |
| singleLine | `is-single-line` | only show a single line of the text. [see here](https://github.com/Automattic/wp-calypso/issues/16933) |

The devdocs postcomments look like this:

```
singleLine
 - excerpt
full
```

To Test: 
1. go to: http://calypso.localhost:3000/devdocs/blocks/post-comment and verify you see PostComments there.  as well as is-single-line, is-full, and is-excerpt comments
2. verify that fullpost still works as expected.

### CODE REVIEW

File: client/blocks/comments/post-comment.jsx [hunk 1]
```diff
@@ -250,7 +290,7 @@ class PostComment extends Component {
 					/> }
 
 				<CommentActions
-					post={ this.props.post }
+					post={ this.props.post || {} }
```
Comment (samouri) [inline client/blocks/comments/post-comment.jsx]: needed to add the default empty post so that CommentActions doesn't barf.

**Comment by samouri** [comment]
added the ReadMore span w/ class `comments__comment-read-more`

**Comment by jancavan** [comment]
@samouri hmm, so I realized that I'm not sure how this will work. 

"Read more" needs to be added immediately after where the text ends. With CSS alone, we can do that by displaying the comment inline, but the problem is inline elements don't allow a width/height so a `max-height` can't be set to limit the number of lines.

It works how it is now (although alignment is still off)...
screenshot 2017-08-16 21 01 46

...but not when comments are short:
screenshot 2017-08-16 21 02 01

(ie if they have line breaks and we're only showing the first line):
screenshot 2017-08-16 21 02 25

**Comment by jancavan** [comment]
@blowery I'm not sure why the "Read more" link has disappeared even now that the text clipping has been put back in place, but this is what it will look like once we have it back:

Single line comment:
screenshot 2017-08-17 19 41 45

Triple line comment:
screenshot 2017-08-17 19 41 50

File: client/blocks/comments/docs/post-comment-example.jsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	'4': {
+		children: [ 1 ],
+		data: mockComments[ 4 ],
+	},
+};
+
+export default class PostCommentExample extends React.Component {
+	static displayName = 'PostCommentExample';
+
+	render() {
```
Comment (samouri) [inline client/blocks/comments/docs/post-comment-example.jsx]: nice examples

File: client/components/auto-direction/index.jsx [hunk 3]
```diff
@@ -171,6 +171,9 @@ const setChildDirection = ( child ) => {
 	if ( child && child.props.children ) {
 		let innerChildDirection = null;
 		const children = React.Children.map( child.props.children, innerChild => {
+			if ( ! innerChild ) {
```
Comment (samouri) [inline client/components/auto-direction/index.jsx]: what is this about?

File: client/blocks/comments/post-comment-content.jsx [hunk 4]
```diff
@@ -21,21 +38,25 @@ export default class PostCommentContent extends React.Component {
...
 		return (
 			<AutoDirection>
-				<div
-					className="comments__comment-content"
-					dangerouslySetInnerHTML={ { __html: this.props.content } }
-				/>
+				<div className={ classNames( 'comments__comment-content-wrapper', this.props.className ) }>
+					<div
+						className="comments__comment-content"
+						ref="overflowNode"
```
Comment (samouri) [inline client/blocks/comments/post-comment-content.jsx]: this method of setting a ref is considered "legacy" as described in the [React Docs](https://facebook.github.io/react/docs/refs-and-the-dom.html#legacy-api-string-refs). 

Lets avoid it!

File: client/components/overflow-watcher/index.jsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+			if ( this.state.overflowX !== overflowX ) {
+				this.setState( { overflowX } );
+			}
+			if ( this.state.overflowY !== overflowY ) {
+				this.setState( { overflowY } );
+			}
+		};
+
+		render() {
+			return <Comp ref={ this.storeRef } { ...this.props } { ...this.state } />;
```
Comment (samouri) [inline client/components/overflow-watcher/index.jsx]: can we forego forcing the child to have a particular ref entirely by wrapping the Component with a reffed div?

If yes, then we might be able to merge this HoC with `withDimensions`.  Right now withDimensions hands down: `{ width, height }`, but could also return `{width, height, overflowY, overflowX }`.  Thats up to you though

**Comment by samouri** [comment]
Functionality LGTM.  Can't wait to get this on conversations!

File: client/components/overflow-watcher/index.jsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		checkSize = () => {
+			if ( ! this.sizeNode ) {
+				this.setState( {
+					overflowX: false,
+					overflowY: false,
+				} );
+			}
+
+			const domNode = findDOMNode( this.sizeNode );
+			const overflowX = domNode.scrollWidth > domNode.clientWidth + 4;
```
Comment (samouri) [inline client/components/overflow-watcher/index.jsx]: what is the 4?

File: client/blocks/comments/post-comment-content.jsx (hunk 4 above)
Comment (blowery) [inline client/blocks/comments/post-comment-content.jsx]: Can't. At least not afaict. The OverflowContainer needs to be able to find the overflow ref in the child. I tried setting a prop on the child via `this.overflowNode = node` in a ref handler, but then the OverflowContainer can't see the nested ref...

File: client/lib/with-dimensions/index.jsx [hunk 7]
```diff
@@ -14,50 +16,59 @@ import { debounce } from 'lodash';
  * @param {object} EnhancedComponent - react component to wrap and give the prop width/height to
  * @returns {object} the enhanced component
  */
-export default EnhancedComponent => class WithWidth extends React.Component {
-	static displayName = `WithDimensions( ${ EnhancedComponent.displayName || EnhancedComponent.name } )`;
-	static propTypes = { domTarget: React.PropTypes.object };
+export default EnhancedComponent =>
```
Comment (samouri) [inline client/lib/with-dimensions/index.jsx]: why the heck did maxWidth change for prettier

File: client/components/overflow-watcher/index.jsx (hunk 5 above)
Comment (blowery) [inline client/components/overflow-watcher/index.jsx]: Yeah, that would be ideal, but the problem is that a component above the thing that's overflowing is what needs the class. So we have
```
<Parent>
  <OverflowingKid />
</Parent>
```
and we need to tell host "your kid is overflowing". I'm not sure how to communicate that back up the tree without using a nested ref.

File: client/components/overflow-watcher/index.jsx (hunk 6 above)
Comment (blowery) [inline client/components/overflow-watcher/index.jsx]: ah, meant to move that to  a const. Just a buffer. I thought I saw some cases where clientHeight and scrollHeight will be different by a pixel. Might just pull it for now and see what happens.

File: client/components/overflow-watcher/index.jsx (hunk 5 above)
Comment (samouri) [inline client/components/overflow-watcher/index.jsx]: figured it out, its not that bad.  not great, but not too bad.  the solution I wrote up could also clean up about half of the other usages of withDimensions

File: client/components/overflow-watcher/index.jsx (hunk 6 above)
Comment (samouri) [inline client/components/overflow-watcher/index.jsx]: pulled it, seemed to work!

File: client/components/auto-direction/index.jsx (hunk 3 above)
Comment (blowery) [inline client/components/auto-direction/index.jsx]: if you have 
```
<AutoDirection>
  <SomeComponent />
  { false && <OptionalComponent />
</AutoDirection>
```
`AutoDirection` gets two children, `SomeComponent` and `undefined`. The check skips undefined children.

Comment (samouri) [inline client/components/auto-direction/index.jsx]: oh that makes sense! thank you for the explanation!

File: client/blocks/comments/post-comment-content.jsx (hunk 4 above)
Comment (blowery) [inline client/blocks/comments/post-comment-content.jsx]: ah, we can pass down a ref setter like with withDimensions.

**Comment by blowery** [comment]
might need to add that buffer back in..

https://www.dropbox.com/s/24kveqp3c01s1yz/Screenshot%202017-08-22%2012.56.43.png?dl=1

**Comment by samouri** [comment]
@blowery: i wonder if the buffer is related to border lines.  i'll throw it back in

**Pull request merged**